ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManagerMetrics.java

```diff
@@ -0,0 +1,242 @@
+
+import com.codahale.metrics.Counter;
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import java.util.List;
+
+
+/**
+ * Metrics for the {@link HelixClusterManager}
+ */
+class HelixClusterManagerMetrics {
+  private final HelixClusterManager.HelixClusterManagerCallback clusterMapCallback;
+  private final MetricRegistry registry;
+
+  public final Counter liveInstanceChangeTriggerCount;
+  public final Counter externalViewChangeTriggerCount;
+  public final Counter instanceConfigChangeTriggerCount;
+  public final Counter getPartitionIdFromStreamMismatchCount;
+  public final Counter getWritablePartitionIdsMismatchCount;
+  public final Counter hasDatacenterMismatchCount;
+  public final Counter getDataNodeIdMismatchCount;
+  public final Counter getReplicaIdsMismatchCount;
+  public final Counter getDataNodeIdsMismatchCount;
+
+  public Gauge<Boolean> helixClusterManagerInstantiationFailed;
+
+  /**
+   * Metrics for the {@link HelixClusterManager}
+   * @param registry The {@link MetricRegistry} associated with the {@link HelixClusterManager}
+   * @param  clusterMapCallback The {@link HelixClusterManager.HelixClusterManagerCallback} used to query information
+   *                            from {@link HelixClusterManager}
+   */
+  HelixClusterManagerMetrics(MetricRegistry registry,
+      final HelixClusterManager.HelixClusterManagerCallback clusterMapCallback) {
+    this.clusterMapCallback = clusterMapCallback;
+    this.registry = registry;
+    liveInstanceChangeTriggerCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "liveInstanceChangeTriggerCount"));
+    externalViewChangeTriggerCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "externalViewChangeTriggerCount"));
+    instanceConfigChangeTriggerCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "instanceConfigChangeTriggerCount"));
+    getPartitionIdFromStreamMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getPartitionIdFromStreamMismatchCount"));
+    getWritablePartitionIdsMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getWritablePartitionIdsMismatchCount"));
+    hasDatacenterMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "hasDatacenterMismatchCount"));
+    getDataNodeIdMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getDataNodeIdMismatchCount"));
+    getReplicaIdsMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getReplicaIdsMismatchCount"));
+    getDataNodeIdsMismatchCount =
+        registry.counter(MetricRegistry.name(HelixClusterManager.class, "getDataNodeIdsMismatchCount"));
+  }
+
+  void initializeInstantiationMetric(final boolean instantiated) {
+    helixClusterManagerInstantiationFailed = new Gauge<Boolean>() {
```

Comment:
why is this a gauge? Once a value is set, this can never change right? Counter seems better.